trajectory_executor/src/trajectory_executor/trajectory_executor.cpp

```diff
@@ -0,0 +1,188 @@
+
+        _cur_traj = std::unique_ptr<cav_msgs::TrajectoryPlan>(new cav_msgs::TrajectoryPlan(msg));
+        ROS_DEBUG_STREAM("Successfully swapped trajectories!");
+
+        // TODO: Force traj emit
+        // TODO: Reset timer state
+    }
+
+    void TrajectoryExecutor::onTrajEmitTick(const ros::TimerEvent& te)
+    {
+        std::unique_lock<std::mutex> lock(_cur_traj_mutex);
+        ROS_DEBUG("TrajectoryExecutor tick start!");
+        
+
+        if (_cur_traj != nullptr) {
+            if (_timesteps_since_last_traj > 0) {
+                _cur_traj = std::unique_ptr<cav_msgs::TrajectoryPlan>(new cav_msgs::TrajectoryPlan(trimFirstPoint(*_cur_traj)));
+            }
+            if (!_cur_traj->trajectory_points.empty()) {
+                std::string control_plugin = _cur_traj->trajectory_points[0].controller_plugin_name;
+                std::map<std::string, ros::Publisher>::iterator it = _traj_publisher_map.find(control_plugin);
+                if (it != _traj_publisher_map.end()) {
+                    ROS_DEBUG("Found match for control plugin %s at point %d in current trajectory!",
+                        control_plugin.c_str(),
+                        _timesteps_since_last_traj);
+                    it->second.publish(*_cur_traj);
+                } else {
+                    std::ostringstream description_builder;
+                    description_builder << "No match found for control plugin " 
+                        << control_plugin << " at point " 
+                        << _timesteps_since_last_traj << " in current trajectory!";
+                    shutdown(description_builder.str());
+                }
+                _timesteps_since_last_traj++;
+            } else {
+                shutdown("Ran out of trajectory data to consume!");
+            }
+        } else {
+            ROS_DEBUG("Awaiting initial trajectory publication...");
+        }
+        ROS_DEBUG("TrajectoryExecutor tick completed succesfully!");
+    }
+
+    void TrajectoryExecutor::shutdown(std::string reason) {
+        ROS_ERROR_STREAM("FATAL ERROR: " << reason);
+        cav_msgs::SystemAlert system_alert;
+        system_alert.type = system_alert.FATAL;
+        system_alert.description = reason;
+        ROS_FATAL("Publishing FATAL error on system alert topic!");
+        _default_nh->publishSystemAlert(system_alert);
+        ros::shutdown();
+
+        exit(1);
+    }
+
+    void TrajectoryExecutor::run()
+    {
+        ROS_DEBUG("Starting operations for TrajectoryExecutor component...");
+        _timer = _timer_nh->createTimer(
+            ros::Duration(ros::Rate(this->_min_traj_publish_tickrate_hz)),
+            &TrajectoryExecutor::onTrajEmitTick, 
+            this);
+
+        ros::AsyncSpinner timer_spinner(1, &_timer_callbacks);
+        timer_spinner.start();
+        ROS_DEBUG("Timer thread started!");
+
+        ros::AsyncSpinner msg_spinner(1, &_msg_callbacks);
+        msg_spinner.start();
```

Comment:
Not sure if we need extra spinner as the same reason for avoiding multiple NodeHandles.